### ISSUE TITLE
How to ignore pre commit/push hooks for commit

### ISSUE THREAD

**Issue by KhizerRehan** [jesseduffield/lazygit#1984]
**Topic**
The project i am working has some Husky hooks are installed like
- pre-commit
- pre-push

How i can commit without running such hooks on every commit.

**Menu:**

<img width="882" alt="f8f7bf2108f956029181a71eaca66b14" src="https://user-images.githubusercontent.com/17727069/172158624-c2126d8a-ff1b-4f8c-8e79-266afacf3066.png">

**Even after running an option does run pre-commit hooks defined in repo:**

https://user-images.githubusercontent.com/17727069/172158720-c12bc25a-1a33-4c94-8fb2-ef904190b43a.mp4

**Your thoughts**
What you have to say about the topic

**Comment by mark2185** [comment]
Just to clarify, it happens even if you're running the mapping under `w - commit changes without pre-commit hook`?

**Comment by la-magra** [comment]
Not OP, but I have probably the same setup.

- Repo with git hooks set via husky https://github.com/typicode/husky 
- Run lazygit
- Stage some changes
- Press `w` and commit
- Hooks still run (`git commit` is missing `-n|--no-verify`?)

This is what appears in the log(`@`)
```
Commit
  git commit -m "WIP"
Git output:
husky: Running pre-commit hook
[STARTED] Preparing lint-staged...
[SUCCESS] Preparing lint-staged...
[STARTED] Running tasks for staged files...
...
```

**Desktop:**
$ lazygit --version
commit=v0.34, build date=2022-04-27T13:28:28Z, build source=binaryRelease, version=0.34, os=linux, arch=amd64

**Comment by la-magra** [comment]
@mark2185 My config file in `~/.config/lazygit/config.yml` is completely empty.

I did a quick test putting the default config into it:
```
$ lazygit -c > ~/.config/lazygit/config.yml
```
And now if I prefix my commits with "WIP" now the skipHooks command works as intended.

Replacing `skipHookPrefix: WIP` by `skipHookPrefix: ""` unfortunately does not allow the command to work since it displays the message `You have not configured a commit message prefix for skipping hooks. Set "git.skipHookPrefix = 'WIP'" in your config`

**Comment by mark2185** [comment]
Great! Now we know what's the cause and we can treat it!

In the meantime, please prefix your commit messages with `WIP` for when you want to skip the pre-commit hooks.

**Comment by la-magra** [comment]
As a minor addendum until this issue is solved.

The following custom command does exactly what's required:

```yaml
customCommands:
  - key: "W"
    command: git commit --no-verify
    context: "global"
    subprocess: true
```

lazygit is just that beautiful.

**Comment by zacgodinez** [comment]
> As a minor addendum until this issue is solved.
> 
> The following custom command does exactly what's required:
> 
> ```yaml
> customCommands:
>   - key: "W"
>     command: git commit --no-verify
>     context: "global"
>     subprocess: true
> ```
> 
> lazygit is just that beautiful.

thank you, added prompt for message also.

`
customCommands:
  - key: "W"
    prompts:
      - type: "input"
        title: "Commit"
        initialValue: ""
    command: git commit -m "{{index .PromptResponses 0}}" --no-verify
    context: "global"
    subprocess: true
`

**Comment by zacgodinez** [comment]
> As a minor addendum until this issue is solved.
> 
> The following custom command does exactly what's required:
> 
> ```yaml
> customCommands:
>   - key: "W"
>     command: git commit --no-verify
>     context: "global"
>     subprocess: true
> ```
> 
> lazygit is just that beautiful.

Thank you, added commit message prompt also.

```
customCommands:
  - key: "W"
    prompts:
      - type: "input"
        title: "Commit"
        initialValue: ""
    command: git commit -m "{{index .PromptResponses 0}}" --no-verify
    context: "global"
    subprocess: true

```

**Comment by lougreenwood** [comment]
How do people handle commits when amending a previously skipped commit (which was skipped with the `skipHookPrefix`)?

**Comment by fabiomcosta** [comment]
Sort of related to this, but when lazygit runs "commit" indirectly, for example when running a "move patch into new commit", even when the current commit has a "WIP" prefix the commit command that runs from that action won't have the "--no-verify" flag, and from my understanding there is no way to currently do that :/

**Comment by sriharishangular** [comment]
Hi, I have husky pre-push hook enabled and it is working fine in vscode - windows, but when I try to push from lazy-git in my WSL terminal (Using Astronvim), it is skipping the pre-push hooks and directly pushing. So I basically have the opposite problem to what is described here. Not sure if this is the correct place to ask this.

**Comment by juru1234** [comment]
Would be really nice to also be able to use this when moving patches in new commits or in the index.

**Comment by digitalmaster** [comment]
> How do people handle commits when amending a previously skipped commit (which was skipped with the `skipHookPrefix`)?

Also wondering about this. My workflow with WIP commits involves a lot of iteration and lazygit makes that super easy except for this part.

**Comment by mark2185** [comment]
@digitalmaster if you're able to build from source, could you please try out the branch [`fix-skip-hook`](https://github.com/jesseduffield/lazygit/tree/fix-skip-hook)? It's just a dirty commit, but it would confirm that we're headed in the right direction if it works for you.

**Comment by digitalmaster** [comment]
@mark2185 Works like a charm. Will use for rest of the day and report back if I see any issues but so far so good. Thank you 🙌

**Comment by mark2185** [comment]
Great! I'll open up a PR soon-ish.

**Comment by mark2185** [comment]
> Any updates on this?

There's a draft PR [here](https://github.com/jesseduffield/lazygit/pull/3532) that I still haven't gotten around to finishing.

Feel free to take over, though! (or anyone else)

If not, I'll get back around to it, eventually.

**Comment by kschweiger** [comment]
Warming up this issue again... 🙉 

I see that this

> ```
> customCommands:
>   - key: "W"
>     prompts:
>       - type: "input"
>         title: "Commit"
>         initialValue: ""
>     command: git commit -m "{{index .PromptResponses 0}}" --no-verify
>     context: "global"
>     subprocess: true
> ```

is in principle a solution but I personally do not really understand why lazygit is so opinionated on the pre-commit issue. If there is a separate command with `w`, why also check of a prefix? 

I'm often working in a codebase where it is pretty much impossible to always solve all pre-commit hooks (something like mypy in a python project). So doing "regular" commits with the `-n` flag is quite common. I do not want any WIP or whatever prefix in there. This probably defeats the purpose of pre-commit hooks in some sense but I cannot imagine I'm the only one with that problem. 

I think it would be nice to just have an option to disable the prefix check (e.g. setting the prefix to an emptry string).

**Comment by kschweiger** [comment]
I have been looking at this again and tried to figure out if I can contribute. 

As I see it, the `--no-verify` is set [here](https://github.com/jesseduffield/lazygit/blob/master/pkg/commands/git_commands/commit.go#L94) purely based in the existance of the prefix in the message string. Considering this, it is totally understandable why it work as it does. 

So if somebody would want to add the option, that lazygit can append the flag without a prefix in the message, changes would need be introduces further up, since multiple levels of calls before ending up in `CommitCmdObj` already operate only on the strings. I think you would need to handle it already [in the `HandlWIPCommitPress`](https://github.com/jesseduffield/lazygit/blob/master/pkg/gui/controllers/helpers/working_tree_helper.go#L142) function and then pass it on down the stack until actually building the commit. 

I would be willing to invest more time into this but since this cannot really be accomplished (as far as I can tell) under the current paradigm (passing only strings down the chain of function calls), I would like to know if the maintainers are open to such changes in the first place.

**Comment by kschweiger** [comment]
> I have been looking at this again and tried to figure out if I can contribute.
> 
> As I see it, the `--no-verify` is set [here](https://github.com/jesseduffield/lazygit/blob/master/pkg/commands/git_commands/commit.go#L94) purely based in the existance of the prefix in the message string. Considering this, it is totally understandable why it work as it does.
> 
> So if somebody would want to add the option, that lazygit can append the flag without a prefix in the message, changes would need be introduces further up, since multiple levels of calls before ending up in `CommitCmdObj` already operate only on the strings. I think you would need to handle it already [in the `HandlWIPCommitPress`](https://github.com/jesseduffield/lazygit/blob/master/pkg/gui/controllers/helpers/working_tree_helper.go#L142) function and then pass it on down the stack until actually building the commit.
> 
> I would be willing to invest more time into this but since this cannot really be accomplished (as far as I can tell) under the current paradigm (passing only strings down the chain of function calls), I would like to know if the maintainers are open to such changes in the first place.

Probably @stefanhaller is the best person to comment on this

**Comment by stefanhaller** [comment]
> As I see it, the `--no-verify` is set [here](https://github.com/jesseduffield/lazygit/blob/master/pkg/commands/git_commands/commit.go#L94) purely based in the existance of the prefix in the message string. Considering this, it is totally understandable why it work as it does.
> 
> So if somebody would want to add the option, that lazygit can append the flag without a prefix in the message, changes would need be introduces further up, since multiple levels of calls before ending up in `CommitCmdObj` already operate only on the strings. I think you would need to handle it already [in the `HandlWIPCommitPress`](https://github.com/jesseduffield/lazygit/blob/master/pkg/gui/controllers/helpers/working_tree_helper.go#L142) function and then pass it on down the stack until actually building the commit.

Passing a bool from the controller down to the CmdObj function is not the issue here, we can totally do that if we need to. An example where we do this already is `BranchCommands.LocalDelete`.

The challenge here is the UX design. We don't want to add new keybindings every time we want to add a command-line flag to an existing command. Some more general thoughts about that topic [here](https://github.com/jesseduffield/lazygit/issues/2849).

You could of course argue that in this case we have the `w` keybinding already, so we might as well improve it, and I guess I wouldn't be opposed. One detail question then is whether it should still prefill the WIP prefix in the subject line; it might be confusing, people might think that if they delete it it will commit without `--no-verify` again.

**Comment by mark2185** [comment]
Also rebasing of commits with `--no-verify` in mind needs to be handled somehow.

**Comment by stefanhaller** [comment]
> Also rebasing of commits with `--no-verify` in mind needs to be handled somehow.

This is unrelated. `git rebase --no-verify` only skips the pre-rebase hook, which can be used for preventing a branch from getting rebased at all in the first place. It is not about skipping the pre-commit hook; git rebase doesn't call this one anyway.

(Which is annoying btw, it can cause conflict resolutions during rebases to not be properly formatted, for example. For that reason we have a pre-push hook at work that runs all the pre-commit checks again for every commit.)

**Comment by kschweiger** [comment]
> The challenge here is the UX design. We don't want to add new keybindings every time we want to add a command-line flag to an existing command. Some more general thoughts about that topic https://github.com/jesseduffield/lazygit/issues/2849.
> 
> You could of course argue that in this case we have the w keybinding already, so we might as well improve it, and I guess I wouldn't be opposed. One detail question then is whether it should still prefill the WIP prefix in the subject line; it might be confusing, people might think that if they delete it it will commit without --no-verify again.

Yes. Using the `w` was my idea. I found that currently you can also add the `no-verify` in der regular commit command by just adding the prefix to the message there. Personally I think it would be better design to decouple the two commands. One runs with `--no-verify` on without. Technically you could then get rid of the prefix altogether but I guess this would break the workflow for some people.

If we want to keep everything working as it does now there are not so many options. My idea would be:
- Prefix is allowed to be an empty string
- The regular commit command checks if the prefix (if set to a none-empty string) is there and enables the flag
- The `w` command prefills the prefix (whatever it is) and enables the flag. I guess you could raise an error/warning or something if an none-empty prefix is deleted to address your concern.
- Keep the "WIP" prefix as default and document that setting the prefix it to an empty string basically decouples the two commands.

**Comment by stefanhaller** [comment]
Sounds good for now, except for some details:

> * Prefix is allowed to be an empty string
> * The regular commit command checks if the prefix (if set to a none-empty string) is there and enables the flag

This is how it behaves already now, so no change needed, or am I missing something?

> * The `w` command prefills the prefix (whatever it is) and enables the flag.

Here's where I differ. I suppose using the WIP prefix for regular commits might be convenient for some people, so we should keep that mechanism (and those who don't want it can set it to an empty string); but the `w` command is now specifically for explicitly enabling `--no-verify` for commits that don't use the prefix, so I would find it confusing and unnecessary to prefill it.

**Comment by stefanhaller** [comment]
> * Keep the "WIP" prefix as default and document that setting the prefix it to an empty string basically decouples the two commands.

They are decoupled either way, no matter what you set the prefix to. We simply have two different, independent ways to skip pre-commit hooks (ok, just one if you set the prefix to empty).

**Comment by kschweiger** [comment]
> This is how it behaves already now, so no change needed, or am I missing something?

The empty string currently raises an error when pressing `w`  but the other one is current behavior.  I was describing what logic would be needed in the [`HandleWIPCommitPress`](https://github.com/jesseduffield/lazygit/blob/master/pkg/gui/controllers/helpers/working_tree_helper.go#L142) and [`HandleCommitPress`](https://github.com/jesseduffield/lazygit/blob/master/pkg/gui/controllers/helpers/working_tree_helper.go#L151C32-L151C49) functions while replicating how it works currently. 

> Here's where I differ. I suppose using the WIP prefix for regular commits might be convenient for some people, so we should keep that mechanism (and those who don't want it can set it to an empty string); but the w command is now specifically for explicitly enabling --no-verify for commits that don't use the prefix, so I would find it confusing and unnecessary to prefill it.

Sure, I have no strong opinion on prefilling the prefix  when using `w`. Don't want to use the prefix anyways 😅 

I think I can start working on PR with this as a starting point and then we can go from there.

**Comment by stefanhaller** [comment]
> The empty string currently raises an error when pressing `w` 

Ah ok, I hadn't realized that.

> I think I can start working on PR with this as a starting point and then we can go from there.

Sounds good!

**Comment by stefanhaller** [comment]
Closing as addressed by #4374.

**Comment by jesseduffield** [comment]
@stefanhaller I just realised that https://github.com/jesseduffield/lazygit/pull/4374 no longer pre-fills the commit message panel with the commit prefix (in my case, 'WIP'). I find this quite annoying because previously I could create a WIP commit with just w + enter, and now I need to type something. I create these WIP's several times a day when I want to store uncommitted changes to my branch before switching to another branch (with the intention of popping them off later, much like using a stash). I really want this flow to remain only a couple keypresses.

If I understand correctly there's now two ways to skip hooks:
1) using the 'w' keybinding
2) using the 'c' keybinding and typing in the commit prefix

What if we just ditched the second option (which was always a bit of a hack) and made it so that the WIP prefix is used when you press 'w' and if the user wants to set that to an empty string, they can. I'm also open to instead having it default to an empty string but I suspect most people only want to skip hooks when committing a WIP.

I notice that in the activity log above that @mikavilpas has just updated created a custom-command to get back the old behaviour too [here](https://github.com/mikavilpas/dotfiles/commit/d4f64107191520382aa0ceaf64c96d1b9fc4fef7) which suggests I'm not the only one feeling this way.

**Comment by stefanhaller** [comment]
I don't have a strong opinion here, so whatever works for you. In the discussion above I had recommended to no longer prefill the commit message with the prefix in order to make things clearer; `w` now always skips hooks independent of the message, so if we still prefill it, people might think that if they press `w` but then delete the prefix, it would commit with hooks again, which it doesn't.

And honestly, if you want a quick way to create "WIP" commits, then a custom command sounds like an appropriate tool to me for this. You could even save a keystroke by binding it to `w` (no enter needed any more).

**Comment by kschweiger** [comment]
The important thing for me was that commits can be created with `--no-verify` and arbitrary commit messages. The changes would haven been necessary anyways if you want to allow an empty string for the prefix. Personally I'm more an @stefanhaller side but as long as I can disable the prefix I'm fine 😀

**Comment by stefanhaller** [comment]
> (and if they do, I'm happy for them to learn that that won't work).

How would they learn though? You can only tell whether or not a hook ran if it fails. There's no way to distinguish "hook succeeded" from "hook didn't run at all", so this could mean they commit some crap that they didn't mean to, and only notice much later when they look back at their history, possibly long after the branch was merged. But I agree that the likelihood for somebody actually doing this is rather low.

> though I don't like how custom commands refresh everything just in case my command changed anything

Why is this a problem? Not a rhetorical question, I'm genuinely curious, because I never noticed. Refreshing is pretty fast and usually doesn't flicker, so in most cases I can't even tell when one happens. And this seems to be unrelated to the `w` command; if you think this is a problem, then maybe it should be addressed in general?

> though I think for now we should just keep the prefix there when you press 'w' to minimise the surprise factor in this release,

That's fine with me, here's the PR: #4456. One small caveat: we do need a small behavior change compared to 0.48, which is that if there is a preserved commit message from an earlier, failed (or aborted) commit, then we use that and not the prefix. This is important for when you try to make a normal commit, the hook fails, and then you want to quickly commit it with `w` but with the same message.

@kschweiger Could you review (and/or test) the PR please?

**Comment by stefanhaller** [comment]
Yes, that's a good idea, and it should also be easy to do. Instead of replacing the "Commit summary" title, I'd probably rather put it on the right side of the frame, similar to the "(ignoring whitespace)" thing in diffs; probably right before the line length. Something like
```
╭─Commit summary──────────────────(hooks disabled)─ 7 ─────╮
```
And if we want to get fancy, we can even update the hint dynamically for normal commits too, as soon as the user typed the skipHooks prefix.

I'll give this a try when I find some time.
